modules/pam_xauth/pam_xauth.c

```diff
@@ -531,18 +541,27 @@ pam_sm_open_session (pam_handle_t *pamh, int flags UNUSED,
 			   cookiefile);
 	}
 
+	/* Get owner and group of the cookiefile to use it*/
+	if (stat(cookiefile, &st) == 0) {
+		uid = st.st_uid;
+		gid = st.st_gid;
+	} else {
+		uid = getuid();
+		gid = getgid();
+	}
```

Comment:
If I remember correctly, the user has full control of the cookiefile location using $XAUTHORITY,
as result, after this change the user would have full control over uid and gid, which might have security implications.

Reply:
Indeed, there is a security implications there. May I suggest to use the cookifile ownership in the case where $XAUTHORITY env isn't used? In that case, the authority file will be searched in the home directory of the requesting user, and therefore, we can trust it without security implications ?
Otherwise, we could used the requesting user to run the `xauth` command using `rpwd->pw_uid` and `rpwd->pw_gid`. 
Something like that was mentioned on this bug: https://github.com/linux-pam/linux-pam/issues/22

Reply:
Users can also symlink the cookiefile in their home directories to point to wherever they like, so stat(cookiefile) cannot be trusted either.  We can probably trust `PAM_RUSER` like pam_wheel does.

Reply:
That is, if `getuid() == geteuid()` and there is no `PAM_RHOST` set, pam_xauth could probably try `PAM_RUSER`.

Reply:
Agreed, I made the change in this way, refactoring and merging commits.